numba_dpex/core/types/range_types.py

```diff
@@ -0,0 +1,267 @@
+            global_range_struct,
+            range_datamodel.get_field_position("dim1"),
+        )
+        ndrange_struct.gdim2 = c.builder.extract_value(
+            global_range_struct,
+            range_datamodel.get_field_position("dim2"),
+        )
+        ndrange_struct.ldim0 = c.builder.extract_value(
+            local_range_struct,
+            range_datamodel.get_field_position("dim0"),
+        )
+        ndrange_struct.ldim1 = c.builder.extract_value(
+            local_range_struct,
+            range_datamodel.get_field_position("dim1"),
+        )
+        ndrange_struct.ldim2 = c.builder.extract_value(
+            local_range_struct,
+            range_datamodel.get_field_position("dim2"),
+        )
+
+    return NativeValue(
+        ndrange_struct._getvalue(), is_error=c.builder.load(is_error_ptr)
+    )
+
+
+@box(RangeType)
+def box_range(typ, val, c):
+    """
+    Convert a native range structure to a Range object.
+    """
+    ret_ptr = cgutils.alloca_once(c.builder, c.pyapi.pyobj)
+    fail_obj = c.pyapi.get_null_object()
+
+    with ExitStack() as stack:
+        range_struct = cgutils.create_struct_proxy(typ)(
+            c.context, c.builder, value=val
+        )
+
+        ndim_obj = c.box(types.int64, range_struct.ndim)
+        with cgutils.early_exit_if_null(c.builder, stack, ndim_obj):
+            c.builder.store(fail_obj, ret_ptr)
+        dim0_obj = c.box(types.int64, range_struct.dim0)
+        with cgutils.early_exit_if_null(c.builder, stack, dim0_obj):
+            c.builder.store(fail_obj, ret_ptr)
+        dim1_obj = c.box(types.int64, range_struct.dim1)
+        with cgutils.early_exit_if_null(c.builder, stack, dim1_obj):
+            c.builder.store(fail_obj, ret_ptr)
+        dim2_obj = c.box(types.int64, range_struct.dim2)
+        with cgutils.early_exit_if_null(c.builder, stack, dim2_obj):
+            c.builder.store(fail_obj, ret_ptr)
+
+        class_obj = c.pyapi.unserialize(c.pyapi.serialize_object(Range))
+        with cgutils.early_exit_if_null(c.builder, stack, class_obj):
+            c.pyapi.decref(ndim_obj)
+            c.pyapi.decref(dim0_obj)
+            c.pyapi.decref(dim1_obj)
+            c.pyapi.decref(dim2_obj)
+            c.builder.store(fail_obj, ret_ptr)
+        # NOTE: The result of this call is not checked as the clean up
+        # has to occur regardless of whether it is successful. If it
+        # fails `res` is set to NULL and a Python exception is set.
+        if typ.ndim == 1:
+            res = c.pyapi.call_function_objargs(class_obj, (dim0_obj,))
+        elif typ.ndim == 2:
+            res = c.pyapi.call_function_objargs(class_obj, (dim0_obj, dim1_obj))
+        elif typ.ndim == 3:
+            res = c.pyapi.call_function_objargs(
+                class_obj, (dim0_obj, dim1_obj, dim2_obj)
+            )
+        else:
+            raise ValueError("Cannot unbox Range instance.")
+
+        c.pyapi.decref(ndim_obj)
```

Comment:
Same question as in sycl event case. Do we need to decref here? They are still part of the return object, so they are being referenced. I may miss some part when the reference it increased one more time. Comment with clarification why do we need to do it will be really helpful with understanding.

Reply:
Your question actually found a small issue. There is no need to box the `ndim` object at all as it is not stored as an attribute of the `Range` type.

However, your question does apply to the other objects: `dim0_obj`, `dim1_obj`, _etc_.

My reasoning is as follows:

The `dim0_obj` is a new `PyObject` that is returned by the call to the `box` method for that type. The type in this case is an `int`, which is a basic immutable type. Thus, when we pass the `dim0_obj` to the constructor call `c.pyapi.call_function_objargs(class_obj, (dim0_obj,))` the PyObject is passed by "value", _i.e.,_ a new `PyObject` that is now owned by the newly created `Range` instance. The original `dim0_obj` therefore still needs to be properly cleaned up and the `decref` ensures it is eventually cleaned up.

P.S. : Full disclosure, your question forced me to reason about what I was doing, so thank you for it. Otherwise, my code was mostly "monkey see, monkey do" from the Numba extension guide.

Reply:
> Your question actually found a small issue. There is no need to box the `ndim` object at all as it is not stored as an attribute of the `Range` type.

Addressed in 38ee434c570ed5c70b543fe5e153fe7621521bb1

Reply:
Thank you!